File: Adaptiv/backend/services/conversation_service.py
"""
Conversation Service for managing persistent conversations and messages
"""

from typing import List, Optional, Dict, Any
from sqlalchemy.orm import Session
from sqlalchemy import desc, and_
from datetime import datetime
import logging

from models.agents import Conversation, ConversationMessage
from models.core import User

logger = logging.getLogger(__name__)

class ConversationService:
    def __init__(self, db: Session):
        self.db = db

    def create_conversation(self, user_id: int, title: Optional[str] = None) -> Conversation:
        """Create a new conversation for a user"""
        try:
            # Don't auto-generate title - let it be set later when first message is sent
            conversation = Conversation(
                user_id=user_id,
                title=title,  # Can be None initially
                is_active=True
            )
            
            self.db.add(conversation)
            self.db.commit()
            self.db.refresh(conversation)
            
            logger.info(f"Created conversation {conversation.id} for user {user_id}")
            return conversation
            
        except Exception as e:
            self.db.rollback()
            logger.error(f"Error creating conversation: {e}")
            raise

    def get_user_conversations(self, user_id: int, limit: int = 50, include_inactive: bool = False) -> List[Conversation]:
        """Get all conversations for a user, ordered by most recent"""
        try:
            query = self.db.query(Conversation).filter(Conversation.user_id == user_id)
            
            if not include_inactive:
                query = query.filter(Conversation.is_active == True)
            
            conversations = query.order_by(desc(Conversation.updated_at)).limit(limit).all()
            
            logger.info(f"Retrieved {len(conversations)} conversations for user {user_id}")
            return conversations
            
        except Exception as e:
            logger.error(f"Error getting user conversations: {e}")
            raise

    def get_conversation(self, conversation_id: int, user_id: int) -> Optional[Conversation]:
        """Get a specific conversation with user validation"""
        try:
            conversation = self.db.query(Conversation).filter(
                and_(
                    Conversation.id == conversation_id,
                    Conversation.user_id == user_id
                )
            ).first()
            
            if conversation:
                logger.info(f"Retrieved conversation {conversation_id} for user {user_id}")
            else:
                logger.warning(f"Conversation {conversation_id} not found for user {user_id}")
            
            return conversation
            
        except Exception as e:
            logger.error(f"Error getting conversation: {e}")
            raise

    def get_conversation_messages(self, conversation_id: int, user_id: int, limit: int = 100) -> List[ConversationMessage]:
        """Get messages for a conversation with user validation"""
        try:
            # First verify the conversation belongs to the user
            conversation = self.get_conversation(conversation_id, user_id)
            if not conversation:
                return []

            messages = self.db.query(ConversationMessage).filter(
                ConversationMessage.conversation_id == conversation_id
            ).order_by(ConversationMessage.created_at).limit(limit).all()
            
            logger.info(f"Retrieved {len(messages)} messages for conversation {conversation_id}")
            return messages
            
        except Exception as e:
            logger.error(f"Error getting conversation messages: {e}")
            raise

    def add_message(
        self, 
        conversation_id: int, 
        user_id: int,
        role: str, 
        content: str, 
        agent_name: Optional[str] = None,
        tools_used: Optional[List[str]] = None
    ) -> ConversationMessage:
        """Add a message to a conversation"""
        try:
            # Verify the conversation belongs to the user
            conversation = self.get_conversation(conversation_id, user_id)
            if not conversation:
                raise ValueError(f"Conversation {conversation_id} not found for user {user_id}")

            message = ConversationMessage(
                conversation_id=conversation_id,
                role=role,
                content=content,
                agent_name=agent_name,
                tools_used=tools_used
            )
            
            self.db.add(message)
            
            # Update conversation's updated_at timestamp
            conversation.updated_at = datetime.utcnow()
            
            self.db.commit()
            self.db.refresh(message)
            
            logger.info(f"Added {role} message to conversation {conversation_id}")
            return message
            
        except Exception as e:
            self.db.rollback()
            logger.error(f"Error adding message: {e}")
            raise

    def update_conversation_title(self, conversation_id: int, user_id: int, title: str) -> Optional[Conversation]:
        """Update a conversation's title"""
        try:
            conversation = self.get_conversation(conversation_id, user_id)
            if not conversation:
                return None

            conversation.title = title
            conversation.updated_at = datetime.utcnow()
            
            self.db.commit()
            self.db.refresh(conversation)
            
            logger.info(f"Updated title for conversation {conversation_id}")
            return conversation
            
        except Exception as e:
            self.db.rollback()
            logger.error(f"Error updating conversation title: {e}")
            raise

    def delete_conversation(self, conversation_id: int, user_id: int) -> bool:
        """Delete a conversation and all its messages"""
        try:
            conversation = self.get_conversation(conversation_id, user_id)
            if not conversation:
                return False

            # Delete all messages first (cascade should handle this, but being explicit)
            self.db.query(ConversationMessage).filter(
                ConversationMessage.conversation_id == conversation_id
            ).delete()
            
            # Delete the conversation
            self.db.delete(conversation)
            self.db.commit()
            
            logger.info(f"Deleted conversation {conversation_id}")
            return True
            
        except Exception as e:
            self.db.rollback()
            logger.error(f"Error deleting conversation: {e}")
            raise

    def archive_conversation(self, conversation_id: int, user_id: int) -> Optional[Conversation]:
        """Archive a conversation (set is_active to False)"""
        try:
            conversation = self.get_conversation(conversation_id, user_id)
            if not conversation:
                return None

            conversation.is_active = False
            conversation.updated_at = datetime.utcnow()
            
            self.db.commit()
            self.db.refresh(conversation)
            
            logger.info(f"Archived conversation {conversation_id}")
            return conversation
            
        except Exception as e:
            self.db.rollback()
            logger.error(f"Error archiving conversation: {e}")
            raise

    def get_conversation_context(self, conversation_id: int, user_id: int, context_limit: int = 10) -> List[Dict[str, Any]]:
        """Get conversation context for LangGraph (last N messages formatted for API)"""
        try:
            messages = self.get_conversation_messages(conversation_id, user_id, limit=context_limit)
            
            # Convert to format expected by LangGraph
            context = []
            for message in messages[-context_limit:]:  # Get last N messages
                context.append({
                    "role": message.role,
                    "content": message.content,
                    "agent_name": message.agent_name,
                    "tools_used": message.tools_used,
                    "created_at": message.created_at.isoformat()
                })
            
            logger.info(f"Retrieved context with {len(context)} messages for conversation {conversation_id}")
            return context
            
        except Exception as e:
            logger.error(f"Error getting conversation context: {e}")
            raise

    def get_conversation_summary(self, conversation_id: int, user_id: int) -> Dict[str, Any]:
        """Get a summary of the conversation (message count, participants, etc.)"""
        try:
            conversation = self.get_conversation(conversation_id, user_id)
            if not conversation:
                return {}

            message_count = self.db.query(ConversationMessage).filter(
                ConversationMessage.conversation_id == conversation_id
            ).count()
            
            # Get unique agents that participated
            agents = self.db.query(ConversationMessage.agent_name).filter(
                and_(
                    ConversationMessage.conversation_id == conversation_id,
                    ConversationMessage.agent_name.isnot(None)
                )
            ).distinct().all()
            
            agent_names = [agent[0] for agent in agents if agent[0]]
            
            summary = {
                "id": conversation.id,
                "title": conversation.title,
                "created_at": conversation.created_at.isoformat(),
                "updated_at": conversation.updated_at.isoformat(),
                "is_active": conversation.is_active,
                "message_count": message_count,
                "participating_agents": agent_names
            }
            
            logger.info(f"Generated summary for conversation {conversation_id}")
            return summary
            
        except Exception as e:
            logger.error(f"Error getting conversation summary: {e}")
            raise

File: Adaptiv/frontend/src/services/conversationService.ts
import api from './api';

// Types
export interface Conversation {
  id: number;
  title?: string;
  created_at: string;
  updated_at: string;
  is_active: boolean;
  message_count?: number;
  participating_agents?: string[];
}

export interface ConversationMessage {
  id: number;
  conversation_id: number;
  role: 'user' | 'assistant';
  content: string;
  agent_name?: string;
  tools_used?: string[];
  created_at: string;
}

export interface ConversationWithMessages {
  conversation: Conversation;
  messages: ConversationMessage[];
}

export interface CreateConversationRequest {
  title?: string;
}

export interface UpdateConversationRequest {
  title: string;
}

export interface CreateMessageRequest {
  role: 'user' | 'assistant';
  content: string;
  agent_name?: string;
  tools_used?: string[];
}

class ConversationService {
  private baseUrl = '/api/conversations';

  /**
   * Create a new conversation
   */
  async createConversation(request: CreateConversationRequest = {}): Promise<Conversation> {
    try {
      const response = await api.post(this.baseUrl, request);
      return response.data;
    } catch (error: any) {
      console.error('Error creating conversation:', error);
      throw new Error(error.response?.data?.detail || 'Failed to create conversation');
    }
  }

  /**
   * Get all conversations for the current user
   */
  async getConversations(limit: number = 50, includeInactive: boolean = false): Promise<Conversation[]> {
    try {
      const response = await api.get(this.baseUrl, {
        params: { limit, include_inactive: includeInactive }
      });
      return response.data;
    } catch (error: any) {
      console.error('Error getting conversations:', error);
      throw new Error(error.response?.data?.detail || 'Failed to get conversations');
    }
  }

  /**
   * Get a specific conversation with its messages
   */
  async getConversation(conversationId: number, messageLimit: number = 100): Promise<ConversationWithMessages> {
    try {
      const response = await api.get(`${this.baseUrl}/${conversationId}`, {
        params: { message_limit: messageLimit }
      });
      return response.data;
    } catch (error: any) {
      console.error('Error getting conversation:', error);
      throw new Error(error.response?.data?.detail || 'Failed to get conversation');
    }
  }

  /**
   * Update a conversation's title
   */
  async updateConversation(conversationId: number, request: UpdateConversationRequest): Promise<Conversation> {
    try {
      const response = await api.put(`${this.baseUrl}/${conversationId}`, request);
      return response.data;
    } catch (error: any) {
      console.error('Error updating conversation:', error);
      throw new Error(error.response?.data?.detail || 'Failed to update conversation');
    }
  }

  /**
   * Delete a conversation
   */
  async deleteConversation(conversationId: number): Promise<void> {
    try {
      await api.delete(`${this.baseUrl}/${conversationId}`);
    } catch (error: any) {
      console.error('Error deleting conversation:', error);
      throw new Error(error.response?.data?.detail || 'Failed to delete conversation');
    }
  }

  /**
   * Archive a conversation
   */
  async archiveConversation(conversationId: number): Promise<Conversation> {
    try {
      const response = await api.post(`${this.baseUrl}/${conversationId}/archive`);
      return response.data;
    } catch (error: any) {
      console.error('Error archiving conversation:', error);
      throw new Error(error.response?.data?.detail || 'Failed to archive conversation');
    }
  }

  /**
   * Add a message to a conversation
   */
  async addMessage(conversationId: number, request: CreateMessageRequest): Promise<ConversationMessage> {
    try {
      const response = await api.post(`${this.baseUrl}/${conversationId}/messages`, request);
      return response.data;
    } catch (error: any) {
      console.error('Error adding message:', error);
      throw new Error(error.response?.data?.detail || 'Failed to add message');
    }
  }

  /**
   * Get conversation context for LangGraph
   */
  async getConversationContext(conversationId: number, contextLimit: number = 10): Promise<any[]> {
    try {
      const response = await api.get(`${this.baseUrl}/${conversationId}/context`, {
        params: { context_limit: contextLimit }
      });
      return response.data.context;
    } catch (error: any) {
      console.error('Error getting conversation context:', error);
      throw new Error(error.response?.data?.detail || 'Failed to get conversation context');
    }
  }

  /**
   * Generate a conversation title based on the first user message
   */
  generateConversationTitle(firstMessage: string): string {
    // Take first 50 characters and add ellipsis if longer
    const maxLength = 50;
    if (firstMessage.length <= maxLength) {
      return firstMessage;
    }
    return firstMessage.substring(0, maxLength).trim() + '...';
  }

  /**
   * Format agent name for display
   */
  formatAgentName(agentName?: string): string {
    if (!agentName) return 'Ada';
    
    const agentDisplayNames: { [key: string]: string } = {
      'pricing_orchestrator': '💼 Pricing Expert',
      'web_researcher': '🔍 Market Researcher',
      'algorithm_selector': '⚙️ Algorithm Specialist',
      'database_agent': '🗄️ Database Specialist'
    };
    
    return agentDisplayNames[agentName] || `🤖 ${agentName}`;
  }

  /**
   * Format tools for display
   */
  formatToolName(toolName: string): string {
    return toolName
      .replace(/_/g, ' ')
      .replace(/\b\w/g, l => l.toUpperCase());
  }

  /**
   * Check if a conversation is recent (within last 24 hours)
   */
  isRecentConversation(conversation: Conversation): boolean {
    const oneDayAgo = new Date();
    oneDayAgo.setDate(oneDayAgo.getDate() - 1);
    return new Date(conversation.updated_at) > oneDayAgo;
  }

  /**
   * Format relative time for display
   */
  formatRelativeTime(dateString: string): string {
    const date = new Date(dateString);
    const now = new Date();
    const diffMs = now.getTime() - date.getTime();
    const diffMins = Math.floor(diffMs / (1000 * 60));
    const diffHours = Math.floor(diffMs / (1000 * 60 * 60));
    const diffDays = Math.floor(diffMs / (1000 * 60 * 60 * 24));

    if (diffMins < 1) {
      return 'Just now';
    } else if (diffMins < 60) {
      return `${diffMins}m ago`;
    } else if (diffHours < 24) {
      return `${diffHours}h ago`;
    } else if (diffDays < 7) {
      return `${diffDays}d ago`;
    } else {
      return date.toLocaleDateString();
    }
  }

  /**
   * Group conversations by date
   */
  groupConversationsByDate(conversations: Conversation[]): { [key: string]: Conversation[] } {
    const groups: { [key: string]: Conversation[] } = {};
    
    conversations.forEach(conversation => {
      const date = new Date(conversation.updated_at);
      const today = new Date();
      const yesterday = new Date();
      yesterday.setDate(yesterday.getDate() - 1);
      
      let groupKey: string;
      
      if (date.toDateString() === today.toDateString()) {
        groupKey = 'Today';
      } else if (date.toDateString() === yesterday.toDateString()) {
        groupKey = 'Yesterday';
      } else {
        groupKey = date.toLocaleDateString();
      }
      
      if (!groups[groupKey]) {
        groups[groupKey] = [];
      }
      groups[groupKey].push(conversation);
    });
    
    return groups;
  }
}

export default new ConversationService();
